--- a/utils/python/timing/neom8t.py
+++ b/utils/python/timing/neom8t.py
@@ -23,7 +23,6 @@ Use USB to write messages and use I2C to
 import os
 import sys
 import time
-import serial
 
 from common.logger import Logger
 from timing.gpsusb import GPSUSB
@@ -78,11 +77,7 @@ class NEOM8T:
         return tpCmd.getAntennaCableDelay()
 
     def setGPSToDTimingFormat(self):
-        ser = serial.Serial()
-        ser.port = "/dev/ttyACM0"
-        ser.baudrate = 9600
-        ser.timeout = 1
-        ser.open()
+        ser = open("/dev/ttyACM0", "wb")
 
         data = ["B56206010800F0020000000000000131",
                 "B56206010800F001000000000001012B",
@@ -97,6 +92,6 @@ class NEOM8T:
 
         for x in data:
             ser.write(bytes.fromhex(x))
+            ser.flush()
 
-        ser.flush()
         ser.close()
